src/Symfony/Component/DependencyInjection/ContainerBuilder.php

```diff
@@ -348,21 +343,11 @@ public function set($id, $service)
 
         if ($this->isFrozen()) {
             // setting a synthetic service on a frozen container is alright
-            if (
-                (!isset($this->definitions[$id]) && !isset($this->obsoleteDefinitions[$id]))
-                    ||
-                (isset($this->definitions[$id]) && !$this->definitions[$id]->isSynthetic())
-                    ||
-                (isset($this->obsoleteDefinitions[$id]) && !$this->obsoleteDefinitions[$id]->isSynthetic())
-            ) {
+            if (!isset($this->definitions[$id]) || !$this->definitions[$id]->isSynthetic()) {
                 throw new BadMethodCallException(sprintf('Setting service "%s" on a frozen container is not allowed.', $id));
```

Comment:
Thanks @ro0NL . I've changed the message to mention it only concerns non-synthetic services.

About your other suggestion, I think it should be considered as a new "DX" feature, as it has been done in the past for similar exceptions giving better hint to the end user.